### PR TITLE
Add support for headers and vcpkg manifest mode

### LINKED ISSUE #23: Replace PMM with `vcpkg.json` generation + bootstrapping
PMM is awfully complicated and does not work properly. Replace it with modern vcpkg.

```toml
[project]
name = "my_project"
version = "1.2"

[vcpkg]
version = "2021.05.12"
packages = ["fmt", "sqlite3"]
```

Should generate `vcpkg.json` (take care of the vcpkg identifiers not allowing capital letters or underscores):

```json
{
  "name": "vcpkg_template",
  "version": "0.0.1",
  "dependencies": ["fmt", "sqlite3"]
}
```

And (right after the `project`):

```cmake
include(FetchContent)

message(STATUS "Fetching vcpkg...")
FetchContent_Declare(
	vcpkg
	URL
		https://github.com/microsoft/vcpkg/archive/refs/tags/2021.05.12.tar.gz
)
FetchContent_MakeAvailable(vcpkg)

include(${vcpkg_SOURCE_DIR}/scripts/buildsystems/vcpkg.cmake)
```

See: https://github.com/build-cpp/vcpkg_template/tree/simplified-vscode. It also needs some kind of handling for subprojects (that scenario is not supported by vcpkg according to one of the authors) and it would be best if you get error messages that certain packages need to be specified in the root project for it to work.

### PR DESCRIPTION
**Pull request by MoAlyousef**
The headers option creates a tree by itself instead of adding to sources, this allows IDE support without having to pass headers to target_sources, AFAIK this prevents internal headers from being treated as PUBLIC_HEADERS.

Closes #23

### CODE REVIEW

File: src/gen.cpp [hunk 1]
```diff
@@ -520,15 +522,26 @@ int generate_cmake(const char *path, bool root) {
...
+        cmd("message")("STATUS", "Fetching vcpkg...");
+        cmd("FetchContent_Declare")("vcpkg", "URL", "https://github.com/microsoft/vcpkg/archive/refs/tags/2021.05.12.tar.gz");
+        cmd("FetchContent_MakeAvailable")("vcpkg");
+        cmd("include")("${vcpkg_SOURCE_DIR}/scripts/buildsystems/vcpkg.cmake");
+        using namespace nlohmann;
+        json j;
+        j["$schema"] = "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json";
+        j["name"] = cmake.project_name;
+        if (cmake.project_version.empty())
+            throw std::runtime_error("vcpkg manifest mode requires that the project have a version string!");
```
Comment (mrexodia) [inline src/gen.cpp:536]: I don't think this is true, for me it was working fine without the version field

File: src/gen.cpp [hunk 2]
```diff
@@ -520,15 +522,26 @@ int generate_cmake(const char *path, bool root) {
...
-        cmd("include")("${pmm_SOURCE_DIR}/pmm.cmake");
-        tsl::ordered_map<std::string, std::vector<std::string>> vcpkg_args;
-        vcpkg_args["REVISION"] = {cmake.vcpkg.version};
-        vcpkg_args["REQUIRES"] = cmake.vcpkg.packages;
-        auto vcpkg = std::make_pair("VCPKG", vcpkg_args);
-        cmd("pmm")(vcpkg).endl();
+    if (!cmake.vcpkg.packages.empty()) {
+        cmd("include")("FetchContent");
+        cmd("message")("STATUS", "Fetching vcpkg...");
+        cmd("FetchContent_Declare")("vcpkg", "URL", "https://github.com/microsoft/vcpkg/archive/refs/tags/2021.05.12.tar.gz");
```
Comment (mrexodia) [inline src/gen.cpp:528]: Probably it would be cleaner to allow:

```toml
[vcpkg]
version = "2021.05.12"
packages = ["..."]
```

And generate the URL from that, or alternatively allow:

```toml
[vcpkg]
url = "https://my-company/vcpkg/blah.tar.gz"
packages = ["..."]
```

That way cmkr doesn't lock the user into a specific vcpkg version.

File: src/gen.cpp [hunk 3]
```diff
@@ -520,15 +522,26 @@ int generate_cmake(const char *path, bool root) {
...
+        cmd("FetchContent_MakeAvailable")("vcpkg");
+        cmd("include")("${vcpkg_SOURCE_DIR}/scripts/buildsystems/vcpkg.cmake");
+        using namespace nlohmann;
+        json j;
+        j["$schema"] = "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json";
+        j["name"] = cmake.project_name;
```
Comment (mrexodia) [inline src/gen.cpp:534]: According to https://vcpkg.readthedocs.io/en/latest/specifications/manifests/#definitions there are a bunch of requirements on the identifier:

- `<identifier>`: A `string` which:
  - Is entirely ASCII
  - Contains only lowercase alphabetic characters, digits, and hyphen-minus
  - Does not have multiple consecutive hyphens
  - Does not begin nor end with a hyphen
  - Is not a Windows filesystem reserved name
  - Is not a vcpkg reserved name: "default" or "core".
  - In other words, it must follow the regex `[a-z0-9]+(-[a-z0-9]+)*`, and must not be any of:
`{ prn, aux, nul, con, lpt[1-9], com[1-9], core, default }`

I would propose a `vcpkg_escape_identifier` function that generates a correct name

File: src/gen.cpp [hunk 4]
```diff
@@ -520,15 +522,26 @@ int generate_cmake(const char *path, bool root) {
...
+        if (cmake.project_version.empty())
+            throw std::runtime_error("vcpkg manifest mode requires that the project have a version string!");
+        j["version"] = cmake.project_version;
+        j["dependencies"] = cmake.vcpkg.packages;
+        std::ofstream ofs("vcpkg.json");
+        if (!ofs) {
+            throw std::runtime_error("Failed to create a vcpkg.json manifest file!");
+        }
+        ofs << std::setw(4) << j << std::endl;
```
Comment (mrexodia) [inline src/gen.cpp:543]: The convention in vcpkg appears to be `setw(2)`, but it's a very minor nitpick

File: src/gen.cpp [hunk 5]
```diff
@@ -520,15 +522,26 @@ int generate_cmake(const char *path, bool root) {
...
+        ofs.close();
```
Comment (mrexodia) [inline src/gen.cpp:544]: No need to close, the destructor does that

File: src/gen.cpp [hunk 6]
```diff
@@ -603,8 +616,21 @@ int generate_cmake(const char *path, bool root) {
                                  [&](const std::string &, const std::vector<std::string> &includes) { inject_includes(includes); });
             gen.handle_condition(target.cmake_before, [&](const std::string &, const std::string &cmake) { inject_cmake(cmake); });
 
+            auto headers_var = target.name + "_HEADERS";
```
Comment (mrexodia) [inline src/gen.cpp:619]: I think the `headers` should just be appended to the `target_SOURCES` variable to actually also add them to `target_sources`. CMake doesn't have a concept of headers and when they are part of the `target_sources` you can get them from cmake etc.

**Review by mrexodia — CHANGES_REQUESTED** [review]
I requested quite a few changes, but if you don't have time/don't feel like doing them I'll work on it in the next couple of days 🙂

File: src/gen.cpp (hunk 1 above)
Comment (MoAlyousef) [reply src/gen.cpp]: Hmm I was going with their docs, I've removed the requirement.

File: src/gen.cpp (hunk 2 above)
Comment (MoAlyousef) [reply src/gen.cpp]: added a url field to Vcpkg.

File: src/gen.cpp (hunk 3 above)
Comment (MoAlyousef) [reply src/gen.cpp]: I see. I've added a check but it generates a temporary name to avoid throwing.

File: src/gen.cpp (hunk 4 above)
Comment (MoAlyousef) [reply src/gen.cpp]: fixed

File: src/gen.cpp (hunk 5 above)
Comment (MoAlyousef) [reply src/gen.cpp]: fixed

File: src/gen.cpp (hunk 6 above)
Comment (MoAlyousef) [reply src/gen.cpp]: I see. The headers are now added to the sources.

File: src/gen.cpp (hunk 1 above)
Comment (mrexodia) [reply src/gen.cpp]: Yeah the docs are right, just tried it and it fails. Currently in the process adding some tests and I'll fix it

Comment (mrexodia) [reply src/gen.cpp]: I found that `"version-string": ""` is working so I'll just unconditionally set that

**Review by mrexodia — APPROVED** [review]

**Comment by mrexodia** [comment]
Awesome, thanks!
